From ea3f864969798ba1290ce7295f7011fb12a5b9b5 Mon Sep 17 00:00:00 2001
From: Paul Keith <javelinanddart@gmail.com>
Date: Wed, 6 Feb 2019 19:09:21 -0600
Subject: [PATCH 1/8] livedisplay: Fix racy init

* It turns out moving this to class main doesn't work as well
  as some had thought it would, so restore the previous logic
  to use class hal, and restart if the backend isn't up yet

Change-Id: Ib8b417d89108dcf80d0d159e8d84e5327944b2a1
---
 legacymm/service.cpp                                        | 5 +++++
 legacymm/vendor.lineage.livedisplay@2.0-service-legacymm.rc | 2 +-
 sdm/lineage.livedisplay@2.0-service-sdm.rc                  | 2 +-
 sdm/service.cpp                                             | 5 +++++
 sdm/vendor.lineage.livedisplay@2.0-service-sdm.rc           | 2 +-
 5 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/legacymm/service.cpp b/legacymm/service.cpp
index dc37076..66d452f 100644
--- a/legacymm/service.cpp
+++ b/legacymm/service.cpp
@@ -92,6 +92,11 @@ int main() {
         goto shutdown;
     }
 
+    if (!cb->isSupported() && !dm->isSupported() && !pa->isSupported()) {
+        // Backend isn't ready yet, so restart and try again
+        goto shutdown;
+    }
+
     configureRpcThreadpool(1, true /*callerWillJoin*/);
 
     if (cb->isSupported()) {
diff --git a/legacymm/vendor.lineage.livedisplay@2.0-service-legacymm.rc b/legacymm/vendor.lineage.livedisplay@2.0-service-legacymm.rc
index 8656ffd..ecb6e6a 100644
--- a/legacymm/vendor.lineage.livedisplay@2.0-service-legacymm.rc
+++ b/legacymm/vendor.lineage.livedisplay@2.0-service-legacymm.rc
@@ -1,4 +1,4 @@
 service vendor.livedisplay-hal-2-0-legacymm /vendor/bin/hw/vendor.lineage.livedisplay@2.0-service-legacymm
-    class main
+    class hal
     user system
     group system
diff --git a/sdm/lineage.livedisplay@2.0-service-sdm.rc b/sdm/lineage.livedisplay@2.0-service-sdm.rc
index ca5eefc..dd76017 100644
--- a/sdm/lineage.livedisplay@2.0-service-sdm.rc
+++ b/sdm/lineage.livedisplay@2.0-service-sdm.rc
@@ -1,4 +1,4 @@
 service livedisplay-hal-2-0-sdm /system/bin/hw/lineage.livedisplay@2.0-service-sdm
-    class main
+    class hal
     user system
     group system
diff --git a/sdm/service.cpp b/sdm/service.cpp
index f561ad9..59a9463 100644
--- a/sdm/service.cpp
+++ b/sdm/service.cpp
@@ -128,6 +128,11 @@ int main() {
         goto shutdown;
     }
 
+    if (!cb->isSupported() && !dm->isSupported() && !pa->isSupported()) {
+        // Backend isn't ready yet, so restart and try again
+        goto shutdown;
+    }
+
     configureRpcThreadpool(1, true /*callerWillJoin*/);
 
     if (ab->isSupported()) {
diff --git a/sdm/vendor.lineage.livedisplay@2.0-service-sdm.rc b/sdm/vendor.lineage.livedisplay@2.0-service-sdm.rc
index 1e66d30..69eacc5 100644
--- a/sdm/vendor.lineage.livedisplay@2.0-service-sdm.rc
+++ b/sdm/vendor.lineage.livedisplay@2.0-service-sdm.rc
@@ -1,4 +1,4 @@
 service vendor.livedisplay-hal-2-0-sdm /vendor/bin/hw/vendor.lineage.livedisplay@2.0-service-sdm
-    class main
+    class hal
     user system
     group system
-- 
2.17.1

